src/GC/GC.jl

```diff
@@ -9,77 +9,112 @@ module GC
 
 using ..C: C
 
-const ENABLED = Ref(true)
-const QUEUE = C.PyPtr[]
+# `ENABLED`: whether or not python GC is enabled, or paused to process later
+const ENABLED = Threads.Atomic{Bool}(true)
+# this event allows us to `wait` in a task until GC is re-enabled
+const ENABLED_EVENT = Threads.Event()
+
+# this is the queue to process pointers for GC (`C.Py_DecRef`)
+const QUEUE = Channel{C.PyPtr}(Inf)
+
+# this is the task which performs GC from thread 1
+const GC_TASK = Ref{Task}()
+
+# This we use in testing to know when our GC is running
+const GC_FINISHED = Threads.Condition()
 
 """
     PythonCall.GC.disable()
 
-Disable the PythonCall garbage collector.
+Disable the PythonCall garbage collector. This should generally not be required.
 
-This means that whenever a Python object owned by Julia is finalized, it is not immediately
-freed but is instead added to a queue of objects to free later when `enable()` is called.
-
-Like most PythonCall functions, you must only call this from the main thread.
 """
 function disable()
     ENABLED[] = false
+    reset(ENABLED_EVENT)
     return
 end
 
 """
     PythonCall.GC.enable()
 
-Re-enable the PythonCall garbage collector.
-
-This frees any Python objects which were finalized while the GC was disabled, and allows
-objects finalized in the future to be freed immediately.
+Re-enable the PythonCall garbage collector. This should generally not be required.
 
-Like most PythonCall functions, you must only call this from the main thread.
 """
 function enable()
     ENABLED[] = true
+    notify(ENABLED_EVENT)
```

Comment:
We could wait for the GC_finished event here to preserve the semantics that this function call only returns after all GC events are processed